### PR TITLE
feat: memory and stack operations

### PR DESCRIPTION
**Pull request by freesig**
- adds store range to memory
- adds load range to memory
- adds free to memory
- adds reserve to stack
- adds store to stack
- adds load to stack

Mostly tests

### CODE REVIEW

**Review by otrho — COMMENTED** [review]
LGTM.  The `free` is interesting, and I'd say unneeded except for however we end up passing memory between programs down the track. :+1:

File: crates/asm-spec/asm.yml [hunk 1]
```diff
@@ -102,6 +102,34 @@ StateRead:
...
+              stack_in: [len]
+            
+            Load:
+              opcode: 0x0C
+              short: LODS
+              description: Load the value at the given stack depth index relative to the bottom.
+              panics:
+                - Index is out of range.
+              stack_in: [index]
+              stack_out: [value]
```
Comment (mitchmindtree) [inline crates/asm-spec/asm.yml:120]: I was gonna say this is functionally equivalent to `DupFrom`, but just noticed `Load` indexes from the bottom of the stack while `DupFrom` is from the top. I can see both being useful.

**Review by mitchmindtree — APPROVED** [review]
The new `Stack` ops kinda turn the stack into a weird memory/stack hybrid thing, but I can't think of a good reason not to. Lgtm :ok_hand: :shipit: 

It looks like this is purely additive (minus some minor changes to state memory shorthand), I might land this so that I can rebase the `develop` branch and take advantage of these new memory range ops when making the `KeyRange` changes and removing `StateMemory` in favour of the flat `temp_memory`.

File: crates/constraint-vm/src/lib.rs [hunk 2]
```diff
@@ -372,6 +375,20 @@ pub fn step_on_temporary(
             memory.store(addr, w)
         }
         asm::Temporary::Load => stack.pop1_push1(|addr| memory.load(addr)),
+        essential_constraint_asm::Temporary::Free => {
```
Comment (supiket) [inline crates/constraint-vm/src/lib.rs:378]:
```suggestion
        asm::Temporary::Free => {
```

File: crates/constraint-vm/src/lib.rs [hunk 3]
```diff
@@ -372,6 +375,20 @@ pub fn step_on_temporary(
...
+            let addr = stack.pop()?;
+            memory.free(addr)
+        }
+        essential_constraint_asm::Temporary::LoadRange => {
```
Comment (supiket) [inline crates/constraint-vm/src/lib.rs:382]:
```suggestion
        asm::Temporary::LoadRange => {
```

File: crates/constraint-vm/src/lib.rs [hunk 4]
```diff
@@ -372,6 +375,20 @@ pub fn step_on_temporary(
...
+            let [addr, size] = stack.pop2()?;
+            let words = memory.load_range(addr, size)?;
+            Ok(stack.extend(words)?)
+        }
+        essential_constraint_asm::Temporary::StoreRange => {
```
Comment (supiket) [inline crates/constraint-vm/src/lib.rs:387]:
```suggestion
        asm::Temporary::StoreRange => {
```

File: crates/constraint-vm/src/memory/tests.rs [hunk 5]
```diff
@@ -16,3 +18,576 @@ fn test_memory_store_load() {
...
+    // Verify memory state hasn't changed
+    assert_eq!(memory.len().unwrap(), 5);
+}
+
+#[test]
+fn test_free_negative_address() {
+    let mut memory = Memory::new();
+    memory.alloc(5).unwrap();
+
+    // Test with negative index (assuming Word is signed)
```
Comment (supiket) [inline crates/constraint-vm/src/memory/tests.rs:119]: Lol
```suggestion
    // Test with negative index
```

File: crates/constraint-vm/src/memory/tests.rs [hunk 6]
```diff
@@ -16,3 +18,576 @@ fn test_memory_store_load() {
...
+
+    // Record initial capacity
+    let initial_capacity = memory.0.capacity();
+
+    // Free most of the memory
+    memory.free(10).unwrap();
+
+    // Verify capacity has been reduced
+    assert!(memory.0.capacity() < initial_capacity);
+}
```
Comment (supiket) [inline crates/constraint-vm/src/memory/tests.rs:179]: Should we check exact values here?

**Review by supiket — APPROVED** [review]
Just nits

**Pull request merged**